# Reference
[1] &nbsp; Guo FB, Ou HY, Zhang CT. ZCURVE: a new system for recognizing protein-coding genes in bacterial and archaeal genomes. Nucleic Acids Res. 2003 Mar 15;31(6):1780-9. doi: 10.1093/nar/gkg254. [[Pubmed]](https://pubmed.ncbi.nlm.nih.gov/12626720/)  

[2] &nbsp; Zhang CT, Zhang R. A nucleotide composition constraint of genome sequences. Comput Biol Chem. 2004 Apr;28(2):149-53. doi: 10.1016/j.compbiolchem.2004.02.002. [[Pubmed]](https://pubmed.ncbi.nlm.nih.gov/15130543/)  

[3] &nbsp; Zhang CT, Gao F, Zhang R. Segmentation algorithm for DNA sequences. Phys Rev E Stat Nonlin Soft Matter Phys. 2005 Oct;72(4 Pt 1):041917. doi: 10.1103/PhysRevE.72.041917. [[Pubmed]](https://pubmed.ncbi.nlm.nih.gov/16383430/)  

[4] &nbsp; Gao F, Zhang CT. GC-Profile: a web-based tool for visualizing and analyzing the variation of GC content in genomic sequences. Nucleic Acids Res. 2006 Jul 1;34(Web Server issue):W686-91. doi: 10.1093/nar/gkl040. [[Pubmed]](https://pubmed.ncbi.nlm.nih.gov/16845098/)  

[5] &nbsp; Zhang R, Zhang CT. A Brief Review: The Z-curve Theory and its Application in Genome Analysis. Curr Genomics. 2014 Apr;15(2):78-94. doi: 10.2174/1389202915999140328162433. [[Pubmed]](https://pubmed.ncbi.nlm.nih.gov/24822026/)  

[6] &nbsp; Hua ZG, Lin Y, Yuan YZ, Yang DC, Wei W, Guo FB. ZCURVE 3.0: identify prokaryotic genes with higher accuracy as well as automatically and accurately select essential genes. Nucleic Acids Res. 2015 Jul 1;43(W1):W85-90. doi: 10.1093/nar/gkv491. Epub 2015 May 14. [[Pubmed]](https://pubmed.ncbi.nlm.nih.gov/25977299/)  

[7] &nbsp; Wang D, Lai FL, Gao F. Ori-Finder 3: a web server for genome-wide prediction of replication origins in Saccharomyces cerevisiae. Brief Bioinform. 2021 May 20;22(3):bbaa182. doi: 10.1093/bib/bbaa182. [[Pubmed]](https://pubmed.ncbi.nlm.nih.gov/34020544/)  

[8] &nbsp; Lai FL, Gao F. GC-Profile 2.0: an extended web server for the prediction and visualization of CpG islands. Bioinformatics. 2022 Mar 4;38(6):1738-1740. doi: 10.1093/bioinformatics/btab864.  [[Pubmed]](https://pubmed.ncbi.nlm.nih.gov/34954794/)  

[9] Yin ZN, Lai FL, Gao F. Unveiling human origins of replication using deep learning: accurate prediction and comprehensive analysis. Brief Bioinform. 2023 Nov 22;25(1):bbad432. doi: 10.1093/bib/bbad432.  [[Pubmed]](https://pubmed.ncbi.nlm.nih.gov/38008420/)  

[10] Geng YQ, Lai FL, Luo H, Gao F. Nmix: a hybrid deep learning model for precise prediction of 2'-O-methylation sites based on multi-feature fusion and ensemble learning. Brief Bioinform. 2024 Sep 23;25(6):bbae601. doi: 10.1093/bib/bbae601. [[Pubmed]](https://pubmed.ncbi.nlm.nih.gov/39550226/)